agreement/events.go

```diff
@@ -975,37 +1006,42 @@ func (e checkpointEvent) AttachConsensusVersion(v ConsensusVersionView) external
 	return e
 }
 
+// This timestamp is assigned to messages that arrive for round R+1 while the current player
+// is still waiting for quorum on R.
+const pipelinedMessageTimestamp = time.Nanosecond
+
+func getTimestampForEvent(eventRound round, d time.Duration, currentRound round, historicalClocks map[round]historicalClock) time.Duration {
+	if eventRound > currentRound {
+		return pipelinedMessageTimestamp
+	}
+	if eventRound == currentRound {
+		return d
+	}
+	if clock, ok := historicalClocks[eventRound]; ok {
+		return clock.Since()
+	}
+	return time.Duration(0)
+}
+
 // AttachValidatedAt looks for a validated proposal or vote inside a
 // payloadVerified or voteVerified messageEvent, and attaches the given time to
 // the proposal's validatedAt field.
-func (e messageEvent) AttachValidatedAt(d time.Duration, currentRound round) messageEvent {
+func (e messageEvent) AttachValidatedAt(d time.Duration, currentRound round, historicalClocks map[round]historicalClock) messageEvent {
 	switch e.T {
 	case payloadVerified:
-		if e.Input.Proposal.Round() > currentRound {
-			e.Input.Proposal.validatedAt = 1
-		} else {
-			e.Input.Proposal.validatedAt = d
-		}
+		e.Input.Proposal.validatedAt = getTimestampForEvent(e.Input.Proposal.Round(), d, currentRound, historicalClocks)
 	case voteVerified:
-		if e.Input.Vote.R.Round > currentRound {
-			e.Input.Vote.validatedAt = 1
-		} else {
-			e.Input.Vote.validatedAt = d
-		}
+		e.Input.Vote.validatedAt = getTimestampForEvent(e.Input.Vote.R.Round, d, currentRound, historicalClocks)
 	}
 	return e
 }
 
 // AttachReceivedAt looks for an unauthenticatedProposal inside a
 // payloadPresent or votePresent messageEvent, and attaches the given
 // time to the proposal's receivedAt field.
-func (e messageEvent) AttachReceivedAt(d time.Duration, currentRound round) messageEvent {
+func (e messageEvent) AttachReceivedAt(d time.Duration, currentRound round, historicalClocks map[round]historicalClock) messageEvent {
```

Comment:
It would be nice to use the same `switch` construction here as in `AttachValidatedAt`.

Also, why are they both exported?